dask/array/routines.py

```diff
@@ -263,6 +271,26 @@ def rot90(m, k=1, axes=(0, 1)):
         return flip(transpose(m, axes_list), axes[1])
 
 
+def _tensordot_rechunk(array, inner_axes, limit=None):
+    """Optimize rechunking for tensordot computation."""
+    if limit is None:
+        limit = parse_bytes(config.get("array.chunk-size"))
+
+    inner_chunks = [array.shape[ax] for ax in inner_axes]
+    while limit < (np.prod(inner_chunks) * array.dtype.itemsize):
+        inner_chunks[np.argmax(inner_chunks)] = np.max(inner_chunks) // 2
```

Comment:
I like this approach of halving the largest.  In the case of an odd number we'll get one tiny chunk.  It mght make more sense to use "ceil division" rather than floor divisions with `int(math.ceil(np.max(inner_chunks) / 2))`

Reply:
Oh yes, that is a good point. Thanks!